File: backend/models/user/userPermissionModel.js
const mongoose = require('mongoose');
const ModelFactory = require('../modelFactory');

/**
 * 用户权限模型
 * 用于管理用户的特殊权限，如商家权限、营养师权限等
 */
const userPermissionSchema = new mongoose.Schema({
  // 用户ID
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true
  },
  
  // 权限类型
  permissionType: {
    type: String,
    enum: ['merchant', 'nutritionist'],
    required: true
  },
  
  // 权限状态
  status: {
    type: String,
    enum: ['pending', 'approved', 'rejected', 'revoked'],
    default: 'pending'
  },
  
  // 授权信息
  grantedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Admin'
  },
  
  grantedAt: {
    type: Date
  },
  
  // 申请信息
  applicationData: {
    // 申请原因
    reason: String,
    // 联系方式
    contactInfo: {
      phone: String,
      email: String,
      wechat: String
    },
    // 相关资质描述
    qualifications: String,
    // 申请时间
    appliedAt: {
      type: Date,
      default: Date.now
    }
  },
  
  // 审核信息
  reviewData: {
    // 审核意见
    reviewComment: String,
    // 审核时间
    reviewedAt: Date,
    // 审核人
    reviewedBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Admin'
    }
  },
  
  // 权限有效期（可选）
  expiresAt: Date,
  
  // 备注
  remark: String
}, {
  timestamps: true
});

// 复合索引
userPermissionSchema.index({ userId: 1, permissionType: 1 }, { unique: true });
userPermissionSchema.index({ status: 1, permissionType: 1 });
userPermissionSchema.index({ grantedBy: 1 });

// 实例方法：授予权限
userPermissionSchema.methods.grant = function(adminId, comment = '') {
  this.status = 'approved';
  this.grantedBy = adminId;
  this.grantedAt = new Date();
  this.reviewData = {
    reviewComment: comment,
    reviewedAt: new Date(),
    reviewedBy: adminId
  };
  return this.save();
};

// 实例方法：拒绝权限
userPermissionSchema.methods.reject = function(adminId, comment = '') {
  this.status = 'rejected';
  this.reviewData = {
    reviewComment: comment,
    reviewedAt: new Date(),
    reviewedBy: adminId
  };
  return this.save();
};

// 实例方法：撤销权限
userPermissionSchema.methods.revoke = function(adminId, comment = '') {
  this.status = 'revoked';
  this.reviewData = {
    reviewComment: comment,
    reviewedAt: new Date(),
    reviewedBy: adminId
  };
  return this.save();
};

// 静态方法：检查用户是否有特定权限
userPermissionSchema.statics.hasPermission = async function(userId, permissionType) {
  const permission = await this.findOne({
    userId,
    permissionType,
    status: 'approved',
    $or: [
      { expiresAt: { $exists: false } },
      { expiresAt: null },
      { expiresAt: { $gt: new Date() } }
    ]
  });
  return !!permission;
};

// 静态方法：获取用户的所有有效权限
userPermissionSchema.statics.getUserPermissions = async function(userId) {
  const permissions = await this.find({
    userId,
    status: 'approved',
    $or: [
      { expiresAt: { $exists: false } },
      { expiresAt: null },
      { expiresAt: { $gt: new Date() } }
    ]
  });
  return permissions.map(p => p.permissionType);
};

// 静态方法：获取待审核的权限申请
userPermissionSchema.statics.getPendingApplications = async function(permissionType = null) {
  const query = { status: 'pending' };
  if (permissionType) {
    query.permissionType = permissionType;
  }
  
  return this.find(query)
    .populate('userId', 'username email profile')
    .sort({ createdAt: -1 });
};

// 静态方法：批量授予权限
userPermissionSchema.statics.batchGrant = async function(permissionIds, adminId, comment = '') {
  const result = await this.updateMany(
    { _id: { $in: permissionIds }, status: 'pending' },
    {
      $set: {
        status: 'approved',
        grantedBy: adminId,
        grantedAt: new Date(),
        'reviewData.reviewComment': comment,
        'reviewData.reviewedAt': new Date(),
        'reviewData.reviewedBy': adminId
      }
    }
  );
  return result;
};

// 使用ModelFactory创建模型
const UserPermission = ModelFactory.createModel('UserPermission', userPermissionSchema);

module.exports = UserPermission;

File: frontend/flutter/lib/features/permission/data/models/user_permission_model.dart
import 'package:json_annotation/json_annotation.dart';

part 'user_permission_model.g.dart';

@JsonSerializable()
class UserPermissionModel {
  @JsonKey(name: '_id')
  final String id;
  
  final String userId;
  final String permissionType;
  final String status;
  final String? grantedBy;
  final DateTime? grantedAt;
  final ApplicationDataModel? applicationData;
  final ReviewDataModel? reviewData;
  final DateTime? expiresAt;
  final String? remark;
  final DateTime createdAt;
  final DateTime updatedAt;

  const UserPermissionModel({
    required this.id,
    required this.userId,
    required this.permissionType,
    required this.status,
    this.grantedBy,
    this.grantedAt,
    this.applicationData,
    this.reviewData,
    this.expiresAt,
    this.remark,
    required this.createdAt,
    required this.updatedAt,
  });

  factory UserPermissionModel.fromJson(Map<String, dynamic> json) =>
      _$UserPermissionModelFromJson(json);

  Map<String, dynamic> toJson() => _$UserPermissionModelToJson(this);
}

@JsonSerializable()
class ApplicationDataModel {
  final String? reason;
  final ContactInfoModel? contactInfo;
  final String? qualifications;
  final DateTime? appliedAt;

  const ApplicationDataModel({
    this.reason,
    this.contactInfo,
    this.qualifications,
    this.appliedAt,
  });

  factory ApplicationDataModel.fromJson(Map<String, dynamic> json) =>
      _$ApplicationDataModelFromJson(json);

  Map<String, dynamic> toJson() => _$ApplicationDataModelToJson(this);
}

@JsonSerializable()
class ContactInfoModel {
  final String? phone;
  final String? email;
  final String? wechat;

  const ContactInfoModel({
    this.phone,
    this.email,
    this.wechat,
  });

  factory ContactInfoModel.fromJson(Map<String, dynamic> json) =>
      _$ContactInfoModelFromJson(json);

  Map<String, dynamic> toJson() => _$ContactInfoModelToJson(this);
}

@JsonSerializable()
class ReviewDataModel {
  final String? reviewComment;
  final DateTime? reviewedAt;
  final String? reviewedBy;

  const ReviewDataModel({
    this.reviewComment,
    this.reviewedAt,
    this.reviewedBy,
  });

  factory ReviewDataModel.fromJson(Map<String, dynamic> json) =>
      _$ReviewDataModelFromJson(json);

  Map<String, dynamic> toJson() => _$ReviewDataModelToJson(this);
}

@JsonSerializable()
class PermissionStatsModel {
  final PermissionTypeStatsModel merchant;
  final PermissionTypeStatsModel nutritionist;

  const PermissionStatsModel({
    required this.merchant,
    required this.nutritionist,
  });

  factory PermissionStatsModel.fromJson(Map<String, dynamic> json) =>
      _$PermissionStatsModelFromJson(json);

  Map<String, dynamic> toJson() => _$PermissionStatsModelToJson(this);
}

@JsonSerializable()
class PermissionTypeStatsModel {
  final int pending;
  final int approved;
  final int rejected;
  final int revoked;

  const PermissionTypeStatsModel({
    required this.pending,
    required this.approved,
    required this.rejected,
    required this.revoked,
  });

  factory PermissionTypeStatsModel.fromJson(Map<String, dynamic> json) =>
      _$PermissionTypeStatsModelFromJson(json);

  Map<String, dynamic> toJson() => _$PermissionTypeStatsModelToJson(this);
}

// 权限申请请求模型
@JsonSerializable()
class PermissionApplicationRequest {
  final String permissionType;
  final String reason;
  final ContactInfoModel? contactInfo;
  final String? qualifications;

  const PermissionApplicationRequest({
    required this.permissionType,
    required this.reason,
    this.contactInfo,
    this.qualifications,
  });

  factory PermissionApplicationRequest.fromJson(Map<String, dynamic> json) =>
      _$PermissionApplicationRequestFromJson(json);

  Map<String, dynamic> toJson() => _$PermissionApplicationRequestToJson(this);
}

// 权限审核请求模型
@JsonSerializable()
class PermissionReviewRequest {
  final String action;
  final String? comment;

  const PermissionReviewRequest({
    required this.action,
    this.comment,
  });

  factory PermissionReviewRequest.fromJson(Map<String, dynamic> json) =>
      _$PermissionReviewRequestFromJson(json);

  Map<String, dynamic> toJson() => _$PermissionReviewRequestToJson(this);
}

// 批量审核请求模型
@JsonSerializable()
class BatchReviewRequest {
  final List<String> permissionIds;
  final String action;
  final String? comment;

  const BatchReviewRequest({
    required this.permissionIds,
    required this.action,
    this.comment,
  });

  factory BatchReviewRequest.fromJson(Map<String, dynamic> json) =>
      _$BatchReviewRequestFromJson(json);

  Map<String, dynamic> toJson() => _$BatchReviewRequestToJson(this);
}

// 枚举定义
enum PermissionType {
  @JsonValue('merchant')
  merchant,
  @JsonValue('nutritionist')
  nutritionist,
}

enum PermissionStatus {
  @JsonValue('pending')
  pending,
  @JsonValue('approved')
  approved,
  @JsonValue('rejected')
  rejected,
  @JsonValue('revoked')
  revoked,
}

enum ReviewAction {
  @JsonValue('approve')
  approve,
  @JsonValue('reject')
  reject,
}

// 扩展方法
extension PermissionTypeExtension on PermissionType {
  String get displayName {
    switch (this) {
      case PermissionType.merchant:
        return '商家权限';
      case PermissionType.nutritionist:
        return '营养师权限';
    }
  }
}

extension PermissionStatusExtension on PermissionStatus {
  String get displayName {
    switch (this) {
      case PermissionStatus.pending:
        return '待审核';
      case PermissionStatus.approved:
        return '已通过';
      case PermissionStatus.rejected:
        return '已拒绝';
      case PermissionStatus.revoked:
        return '已撤销';
    }
  }
}

extension ReviewActionExtension on ReviewAction {
  String get displayName {
    switch (this) {
      case ReviewAction.approve:
        return '通过';
      case ReviewAction.reject:
        return '拒绝';
    }
  }
}